src/main/java/com/hierynomus/smbj/common/MessageSigning.java

```diff
@@ -0,0 +1,108 @@
+     *            byte array containing the SMB message.
+     * @param len
+     *            message length. Note that the buffer array might contain more
+     *            bytes than the message length. The len parameter indicates how
+     *            many bytes of the buffer array are in the message.
+     * @param signingKey
+     *            the session's signing key.
+     * @return
+     */
+    public static boolean validateSignature(byte[] buffer, int len, byte[] signingKey) {
+        try {
+            byte[] signature = computeSignature(buffer, len, signingKey);
+
+            // are signatures identical?
+            for (int i = 0; i < SMB2Header.SIGNATURE_SIZE; i++) {
+                if (signature[i] != buffer[SMB2Header.SIGNATURE_OFFSET + i])
+                    return false;
+            }
+            return true;
+        } catch (NoSuchAlgorithmException | InvalidKeyException e) {
+            return false; // cannot check signature?
+        }
+    }
+
+    /**
+     * sign a SMB message buffer
+     * 
+     * @param buffer
+     *            byte array containing the SMB message.
+     * @param len
+     *            message length. Note that the buffer array might contain more
+     *            bytes than the message length. The len parameter indicates how
+     *            many bytes of the buffer array are in the message.
+     * @param signingKey
+     *            the session's signing key.
+     * @throws InvalidKeyException
+     * @throws NoSuchAlgorithmException
+     */
+    public static void signBuffer(byte[] buffer, int len, byte[] signingKey) throws InvalidKeyException,
+                    NoSuchAlgorithmException {
+        byte[] signature = computeSignature(buffer, len, signingKey);
+        System.arraycopy(signature, 0, buffer, SMB2Header.SIGNATURE_OFFSET, 16);
+    }
+
+    /**
+     * compute the HMAC signature on a SMB buffer. The calculation is performed
+     * <i>as if the signature field is filled with zeros</i>.
+     * 
+     * @param buffer
+     *            byte array containing the SMB message.
+     * @param len
+     *            message length. Note that the buffer array might contain more
+     *            bytes than the message length. The len parameter indicates how
+     *            many bytes of the buffer array are in the message.
+     * @param signingKey
+     *            the session's signing key.
+     * @throws InvalidKeyException
+     * @throws NoSuchAlgorithmException
+     */
+    private static byte[] computeSignature(byte[] buffer, int len, byte[] signingKey) throws NoSuchAlgorithmException,
+                    InvalidKeyException {
+        if (len < SMB2Header.STRUCTURE_SIZE)
+            throw new IllegalArgumentException("Buffer must be longer than 64 bytes");
+        SecretKeySpec signingKeySpec = new SecretKeySpec(signingKey, MessageSigning.HMAC_SHA256_ALGORITHM);
+        Mac mac = Mac.getInstance(MessageSigning.HMAC_SHA256_ALGORITHM);
```

Comment:
What benefit are you thinking about?  I haven't looked at what is required to do SMBv3 signing, so I can't say if this would be sufficient to avoid having to rewrite the signing code.  I kind of assumed I would have to write a MessageSigningv3 class or something.